opsdroid/connector/twitch/__init__.py

```diff
@@ -0,0 +1,769 @@
+                TWITCH_WEBHOOK_ENDPOINT, headers=headers, json=payload
+            )
+
+            if response.status >= 400:
+                _LOGGER.debug(_("Error: %s - %s"), response.status, response.text)
+
+    async def handle_challenge(self, request):
+        """Challenge handler for get request made by Twitch.
+
+        Upon subscription to a Twitch webhook, Twitch will do a get request to the
+        ``callback`` url provided to check if the url exists. Twitch will do a get request
+        with a challenge and expects the ``callback`` url to return that challenge in plain-text
+        back to Twitch.
+
+        This is what we are doing here, we are getting ``hub.challenge`` from the request and return
+        it in plain-text, if we can't find that challenge we will return a status code 500.
+
+        Args:
+            request (aiohttp.web.Request): Request made to the get route created for webhook subscription.
+
+        Returns:
+           aiohttp.web.Response: if request contains ``hub.challenge`` we return it, otherwise return status 500.
+
+        """
+        challenge = request.rel_url.query.get("hub.challenge")
+
+        if challenge:
+            return aiohttp.web.Response(text=challenge)
+
+        _LOGGER.debug(_("Failed to get challenge from GET Request made by Twitch."))
+        return aiohttp.web.Response(status=500)
+
+    async def twitch_webhook_handler(self, request):
+        """Handle event from Twitch webhooks.
+
+        This method will handle events when they are pushed to the webhook post route. Each webhook will
+        send a different kind of payload so we can handle each event and trigger the right opsdroid event
+        for the received payload.
+
+        For follow events the payload will contain ``from_id`` (broadcaster id), ``from_username`` (broadcaster username)
+        ``to_id`` (follower id), ``to_name`` (follower name) and ``followed_at`` (timestamp).
+
+        For stream changes a lot more things are returned but we only really care about ``type`` (if live/offline)
+        ``title`` (stream title).
+
+        For subscriptions events we will want to know ``event_type`` , ``timestamp`` , ``event_data.plan_name`` , ``event_data.is_gift`` ,
+        ``event_data.tier`` , ``event_data.username`` and ``event_data.gifter_name``.
+
+        Args:
+            request (aiohttp.web.Request): Request made to the post route created for webhook subscription.
+
+        Return:
+            aiohttp.web.Response: Send a ``received`` message and status 200 - Twitch will keep sending the event if it doesn't get the 200 status code.
+
+        """
+
+        valid = await self.validate_request(request, self.webhook_secret)
+        payload = await request.json()
+
+        if valid:
+            try:
+                [data] = payload.get("data")
+                _LOGGER.debug(_("Got event from Twitch - %s") % data)
+
+                if data.get("followed_at"):
```

Comment:
have you considered using our handy [`EventCreator`](https://github.com/opsdroid/opsdroid/blob/master/opsdroid/events.py#L19) to replace this list of ifs?

Reply:
Again I didn't even remember about that but yeah that would be a better way to do it haha I am going to have a look on how you use the EventCreator in matrix and check how to update that huge list of ifs 😁

Reply:
we also use it in slack as well (although it's a little more hacky there).

Reply:
I'm never sure how I feel about the `EventCreator`. It introduces a large amount of [indirection](https://matthewrocklin.com/blog/work/2019/06/23/avoid-indirection) and I often struggle debugging in that area because you have to track so many function calls and pointers in your head.

Reply:
What do you think it might be the best option here? I think on the stream started we need to keep it as it is since I am triggering the listening method on when the user started streaming 🤔  

I guess I could add one for the follow and sub events 🤔

Reply:
At it's core the event creator is a dictionary of `type: function` with a helper to know where to read type from. I do agree that the slack implementation isn't great due to the way the slack API thing works, but by the time we got there we were pretty done with that PR lol.

I am interested in better ways we could do that, I personally have a pretty strong dislike of a couple of hundred lines of if statements.

Reply:
> I personally have a pretty strong dislike of a couple of hundred lines of if statements

Yup I also agree with this. Hence why I'm on the fence about patterns like this.

Reply:
I would potentially argue that the horrible long if statement is very readable compared to the event creator. And making things accessible to beginner developers is a core principal of opsdroid.

But I'd love to find a third option.

Reply:
I took a stab at EventCreator but doesn't seem like a great way to use in this case. One of the biggest issues with it is the `dispatch_key` - Twitch will not use a single key that we can use to create events from,  I've tried sending data as it is, the content of the list inside data and even tried to add a default key and use that, but it didn't work.

I also got to be honest and say that I don't fully understand the EventCreator code. When trying to create the event just from the `UserFollowed` and passing the `followed_at` key as `dispatch_key` it seems that this event was never created anyway.

If I figure out how to use the EventCreator we could potentially reduce the amount of if when receiving a subscriber event - but the other events will either have to stay as they are or create another event creator to handle the other types which seems it would be just double work imo.

Reply:
fair enough.